internal/cri/server/images/image_pull.go

```diff
@@ -108,14 +108,14 @@ func (c *GRPCCRIImageService) PullImage(ctx context.Context, r *runtime.PullImag
 		return ParseAuth(hostauth, host)
 	}
 
-	ref, err := c.CRIImageService.PullImage(ctx, imageRef, credentials, r.SandboxConfig)
+	ref, err := c.CRIImageService.PullImage(ctx, imageRef, credentials, r.SandboxConfig, r.GetImage().GetRuntimeHandler())
```

Comment:
need to normalize how we find the handler / which we are looking for.. the label in sandbox config or the handler passed .. probably the later if also the former just add the label to the sandboxconfig.. if the later and the image handler is nil  can probably specify the handler from the label.. again if it's there.  Either way need to pass the normalized handler all the way through c.snapshotterFromPodSandboxConfig() below so it finds the right snapshotter

Reply:
also.. because our config toml is modifiable.. we might need to store the runtime handler and the snapshotter (labels probably).. for later use wrt matching. See the getLabels() api for the pattern to use there where we add a pinned flag. Though a failure to match will just result in pulling the image again.

Note that we would have to assume no kubelet labels being passed through CRI is means default runtime and snapshotter... until we get that populated in image meta store.  And default can change so we just can't rely on a guess :-O 

TODO: Additionally through ctr and the containerd client we can pull and push images into the k8s.io namespace and that will need some noodling as well to make sure they have a handler / snapshotter option and that we tie the events we receive on said pull pushes to the CRI meta store... All that said if we have handler(s) and/or snapshotter(s) in in the lower level image meta then we don't need it in CRI image store..

Reply:
I've addressed how we read runtimeHandler in the changes for image pull per runtime class changes (was only trying to isolate changes that need not be part of the bigger PR). This PR is only to pass the value of runtimeHandler now that k8s cri-api is at 0.29.1 that has the new runtime handler field that was added through KEP 4216.

We should honor the runtimeHandler if one was specified through CRI and if not, look into the pod annotations to see if one was specified. The latter should be removed once we go to GA in k8s for kep 4216. Since the pod annotations route for specifying the runtime handler is still actively being used, I think we should support that till kep 4216 goes GA.

Also, Derek already added a way to store the runtimehandler and snapshotter per runtime class with his refactoring https://github.com/containerd/containerd/blob/32169d591dbc6133ef7411329b29d0c0433f8c4d/internal/cri/config/config.go#L286 . So we should be good to just rely on this one as only CRIImageService needs to know about resolution from runtime handler to platform/snapshotter.

The intent of this PR is to just merely pass the new runtime handler flag that was added to k8s cri v0.29.1

Reply:
https://github.com/containerd/containerd/blob/main/internal/cri/server/images/image_pull.go#L773

Reply:
i'm not seeing where we add labels(annotations) to the pod config or image... not seeing how we'll remember what we did just yet... maybe a chat

Reply:
@mikebrow yes, this will need to be tweaked and the PR for image pull per runtime will address this. The main intent of this PR is to just to bring in the new runtime handler field.

Reply:
right ok then per discussion will table the use of the runtime handler param to the next PR

Reply:
Has a quick huddle with Mike. Added a todo to the code and will be followed up with the actual PR for kep 4216.